storage/provider/openstack.go

```diff
@@ -0,0 +1,328 @@
+
+	region, _ := cfg.ValueString("region")
+
+	authClient := client.NewClient(&identity.Credentials{
+		URL:        cfg.Attrs()[openstackCfgKeyUrl].(string),
+		User:       cfg.Attrs()[openstackCfgKeyUser].(string),
+		Secrets:    cfg.Attrs()[openstackCfgKeySecrets].(string),
+		Region:     region,
+		TenantName: cfg.Attrs()[openstackCfgKeyTenantName].(string),
+	}, identity.AuthUserPass, nil)
+
+	endpoint := authClient.EndpointsForRegion(region)["volume"]
+	endpointUrl, err := url.Parse(endpoint)
+	if err != nil {
+		return nil, errors.Annotate(err, "error parsing endpoint")
+	}
+	cinderClient := cinder.Basic(endpointUrl, authClient.TenantId(), authClient.Token)
+
+	novaClient := nova.New(authClient)
+
+	return &gooseAdapter{cinderClient, novaClient}, nil
+}
+
+type gooseAdapter struct {
+	cinder *cinder.Client
+	nova   *nova.Client
+}
+
+func (ga *gooseAdapter) CreateVolume(size uint64, name string) (storage.Volume, error) {
+
+	resp, err := ga.cinder.CreateVolume(cinder.CreateVolumeVolumeParams{
+		// NOTE: The Cinder documentation incorrectly states the
+		// size parameter is in GB. It is actually GiB.
+		Size: int(math.Ceil(float64(size / 1024))),
+		Name: name,
+	})
+	if err != nil {
+		return storage.Volume{}, err
+	}
+	return cinderToJujuVolume(&resp.Volume), err
+}
+
+func (ga *gooseAdapter) DeleteVolume(volId string) error {
+	return ga.cinder.DeleteVolume(volId)
+}
+
+func (ga *gooseAdapter) GetVolumesSimple() ([]storage.Volume, error) {
+
+	resp, err := ga.cinder.GetVolumesSimple()
+	if err != nil {
+		return nil, err
+	}
+
+	vols := make([]storage.Volume, len(resp.Volumes))
+	for idx, vol := range resp.Volumes {
+		vols[idx] = cinderToJujuVolume(&vol)
+	}
+
+	return vols, nil
+}
+
+func (ga *gooseAdapter) AttachVolume(serverId, volumeId, mountPoint string) (storage.VolumeAttachment, error) {
+	attachment, err := ga.nova.AttachVolume(serverId, volumeId, mountPoint)
+	return storage.VolumeAttachment{
+		Volume:     names.NewVolumeTag(attachment.VolumeId),
+		Machine:    names.NewMachineTag(attachment.ServerId),
+		DeviceName: attachment.Device,
+	}, err
+}
+
+const (
+	openstackCfgKeyUrl        = "auth-url"
+	openstackCfgKeyUser       = "user"
+	openstackCfgKeySecrets    = "secrets"
+	openstackCfgKeyRegion     = "region-name"
+	openstackCfgKeyTenantName = "tenant-name"
+)
+
+func NewOpenstackStorageConfig(url, user, secrets, region, tenantName string) (*storage.Config, error) {
+
+	return storage.NewConfig(
+		"Openstack",
+		OpenstackProviderType,
+		map[string]interface{}{
+			openstackCfgKeyUrl:        url,
+			openstackCfgKeyUser:       user,
+			openstackCfgKeySecrets:    secrets,
+			openstackCfgKeyRegion:     region,
+			openstackCfgKeyTenantName: tenantName,
+		},
+	)
+}
+
+func cinderToJujuVolume(volume *cinder.Volume) storage.Volume {
+	var size uint64
+	size = uint64(int64(volume.Size * 1024))
+	return storage.Volume{
+		VolumeId: volume.ID,
+		Size:     size,
+		Tag:      names.NewVolumeTag(volume.ID),
```

Comment:
Yeah, the DescribeVolumes signature is broken. It should be returning a different structure, one that does not include a tag; just info that the provider can supply. Passing in an empty tag is a reasonable thing to do for now, as long as there's a TODO to fix it once the API is updated.